File: bin/reddit-download-all.rb
#!/usr/bin/env ruby

require 'open-uri'
require 'json'

usernames = ARGV.dup

def download(url)
  puts url
  raw_data = open(url, "User-Agent" => "Ruby/#{RUBY_VERSION}", &:read)
  [raw_data, JSON.parse(raw_data)]
end

def filename(num)
  num.to_s.rjust(5, "0") + ".json"
end

url = nil
usernames.each do |username|
  num = 0
  base = "https://www.reddit.com/user/" + username + ".json"
  url ||= base
  raw, data = download(url)
  File.write(filename(num), raw)

  while after = data['data']['after']
    sleep(rand(1..10))
    url = base + "?count=25&after=#{after}"
    raw, data = download(url)
    num += 1
    File.write(filename(num), raw)
  end
end

File: bin/reddit-download-all.py
#!/usr/bin/env python

import argparse
import itertools
import json
import random
import time
from pathlib import Path

import requests

parser = argparse.ArgumentParser()
parser.add_argument("usernames", nargs="+", help="the username")
args = parser.parse_args()

headers = {"User-agent": "jtprince_special_sauce 1.23.4"}


def make_filename(num):
    return str(num).zfill(5) + ".pp.json"


def make_placeholder_params(after=None):
    return {} if after is None else dict(count=25, after=after)


for username in args.usernames:
    url = f"https://www.reddit.com/user/{username}.json"
    after = None
    for index in itertools.count(start=1):
        params = make_placeholder_params(after)
        print("params:", params)
        response = requests.get(url, params=params, headers=headers)
        if response.ok:
            data = response.json()
            after = data["data"]["after"]
            if after is None:
                break
            filename = make_filename(index)
            print("Writing to:", filename)
            Path(filename).write_text(json.dumps(data, indent=2))
            sleep_seconds = random.randint(1, 10)
            print("sleeping for:", sleep_seconds)
            time.sleep(sleep_seconds)
        else:
            print("FAILED CALL!")
            print(response.status_code)
            print(response.text)
            print("index:", index)
            print("after:", after)
            breakpoint()
            print("here")
